### PR TITLE
feat: support remote storage (fsspec) in checkpoint consolidation

### LINKED ISSUE #21825: Support remote/cloud storage (fsspec) for checkpoint consolidation
### Description & Motivation

python -m lightning.fabric.utilities.consolidate_checkpoint (and the lightning consolidate CLI) currently only supports consolidating FSDP-sharded distributed checkpoints that live on the local filesystem. _process_cli_args builds a pathlib. Path from the input folder and calls .exists() / .is_dir() / .is_file() on it, and the final consolidated checkpoint is written with a plain torch.save. Both of these assumptions break for checkpoints stored in remote/object storage (S3, GCS, Azure Blob, etc.), which have no real directory semantics and can't be addressed with pathlib.

Lightning already added fsspec URL support for saving/loading distributed checkpoints in ModelParallelStrategy and for FSDP checkpoints more broadly. Checkpoint consolidation is the remaining gap: users training with sharded/FSDP checkpoints on remote storage have no way to consolidate them into a single file without first downloading the shards locally.

### Pitch

_No response_

### Alternatives

_No response_

### Additional context

_No response_

cc @lantiga

### PR DESCRIPTION
**Pull request by deependujha**
## What does this PR do?

Fixes #21825

✅ verified with:

`python -m lightning.fabric.utilities.consolidate_checkpoint s3://lightning-checkpoints/meow-meow-checkpoint-2/lightning_logs/version_0/checkpoints/epoch=1-step=2.ckpt/ --output_file OUTPUT_FILE_DEST`.

both local & remote `output_file_dest` work fine.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/fabric/utilities/consolidate_checkpoint.py [hunk 1]
```diff
@@ -76,5 +81,7 @@ def _process_cli_args(args: Namespace) -> Namespace:
 if __name__ == "__main__":
     args = _parse_cli_args()
     config = _process_cli_args(args)
+    print(f"{config=}")
+    # exit(0)
```
Comment (deependujha) [inline src/lightning/fabric/utilities/consolidate_checkpoint.py]:
```suggestion
```

**Comment by deependujha** [comment]
✅ verified with:

`python -m lightning.fabric.utilities.consolidate_checkpoint s3://lightning-checkpoints/meow-meow-checkpoint-2/lightning_logs/version_0/checkpoints/epoch=1-step=2.ckpt/ --output_file OUTPUT_FILE_DEST`.

both local & remote `output_file_dest` work fine.

File: src/lightning/fabric/utilities/consolidate_checkpoint.py [hunk 2]
```diff
@@ -76,5 +80,7 @@ def _process_cli_args(args: Namespace) -> Namespace:
 if __name__ == "__main__":
     args = _parse_cli_args()
     config = _process_cli_args(args)
+    print(f"{config=}")
+    exit(0)
```
Comment (deependujha) [inline src/lightning/fabric/utilities/consolidate_checkpoint.py]:
```suggestion
```

**Comment by deependujha** [comment]
Hi @zhixiangli, it'll be nice if you can review it when you've some time.

File: src/lightning/fabric/utilities/consolidate_checkpoint.py [hunk 3]
```diff
@@ -77,4 +83,4 @@ def _process_cli_args(args: Namespace) -> Namespace:
     args = _parse_cli_args()
     config = _process_cli_args(args)
     checkpoint = _load_distributed_checkpoint(config.checkpoint_folder)
-    torch.save(checkpoint, config.output_file)
+    _atomic_save(checkpoint, config.output_file)
```
Comment (zhixiangli) [inline src/lightning/fabric/utilities/consolidate_checkpoint.py]: Could you take a look at https://github.com/Lightning-AI/pytorch-lightning/pull/21799 when you have a moment? The switch from torch.save to _atomic_save can cause consolidation (and other similar operations) to silently report success while writing nothing due to a permission error.

File: src/lightning/fabric/utilities/consolidate_checkpoint.py [hunk 4]
```diff
@@ -43,27 +42,34 @@ def _process_cli_args(args: Namespace) -> Namespace:
...
-        _log.error(f"The provided checkpoint folder does not exist: {checkpoint_folder}")
-        sys.exit(1)
-    if not checkpoint_folder.is_dir():
-        _log.error(
-            f"The provided checkpoint path must be a folder, containing the checkpoint shards: {checkpoint_folder}"
-        )
-        sys.exit(1)
-    if not (checkpoint_folder / _METADATA_FILENAME).is_file():
+    checkpoint_folder = _resolve_path(args.checkpoint_folder)
+    checkpoint_folder = checkpoint_folder.expanduser() if isinstance(checkpoint_folder, Path) else checkpoint_folder
```
Comment (zhixiangli) [inline src/lightning/fabric/utilities/consolidate_checkpoint.py]: Should we call expanduser first, and then _resolve_path? It might be good to fold this logic into _resolve_path.

File: src/lightning/fabric/utilities/consolidate_checkpoint.py [hunk 5]
```diff
@@ -43,27 +42,34 @@ def _process_cli_args(args: Namespace) -> Namespace:
...
         sys.exit(1)
 
     if args.output_file is None:
-        output_file = checkpoint_folder.with_suffix(checkpoint_folder.suffix + ".consolidated")
+        output_file = _resolve_path(str(checkpoint_folder).rstrip("/") + ".consolidated")
     else:
-        output_file = Path(args.output_file)
-    if output_file.exists():
+        output_file = _resolve_path(args.output_file)
+    output_file = output_file.expanduser() if isinstance(output_file, Path) else output_file
```
Comment (zhixiangli) [inline src/lightning/fabric/utilities/consolidate_checkpoint.py]: Should we call expanduser first, and then _resolve_path? It might be good to fold this logic into _resolve_path.

Comment (deependujha) [reply src/lightning/fabric/utilities/consolidate_checkpoint.py]: yeah, having this in `resolve_path` makes more sense. Thanks. I'll do it.

File: src/lightning/fabric/utilities/consolidate_checkpoint.py (hunk 3 above)
Comment (deependujha) [reply src/lightning/fabric/utilities/consolidate_checkpoint.py]: interesting, thanks for pointing this out. I'll also try to get #21799 landed soon.

**Review by tchaton — APPROVED** [review]
